include/api/CPerPartitionCategoryIdMapper.h

```diff
@@ -0,0 +1,80 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+#ifndef INCLUDED_ml_api_CPerPartitionCategoryIdMapper_h
+#define INCLUDED_ml_api_CPerPartitionCategoryIdMapper_h
+
+#include <api/CCategoryIdMapper.h>
+#include <api/ImportExport.h>
+
+#include <map>
+#include <vector>
+
+namespace ml {
+namespace api {
+
+//! \brief
+//! Category ID mapper for use with per-partition categorization.
+//!
+//! DESCRIPTION:\n
+//! Maps between global category ID and the tuple
+//! (partition field value, local category ID).
+//!
+//! IMPLEMENTATION DECISIONS:\n
+//! This mapper is designed for jobs that do per-partition
+//! categorization.
+//!
+//! A multi-index stores the mapping so that it can work
+//! both ways without data duplication.
+//!
+//! Although we expect global IDs to run continuously from
+//! 1 to the maximum category ID, the multi-index uses an
+//! ordered tree rather than random access for the global
+//! ID index.  This means that if we decide to completely
+//! discard partitions in the future then the data structure
+//! and persistence mechanism will support this.  For the
+//! same reason the highest ever global ID is stored
+//! separately instead of being obtained from the end of
+//! the global ID index.
+//!
+//! This class is not thread-safe.  Each object should only
+//! be used within a single thread.
+//!
+class API_EXPORT CPerPartitionCategoryIdMapper : public CCategoryIdMapper {
+public:
+    //! Map from a categorizer key and category ID local to that categorizer to
+    //! a global category ID.  This method is not const, as it will create a
+    //! new global ID if one does not exist.
+    CGlobalCategoryId map(const std::string& categorizerKey,
+                          model::CLocalCategoryId localCategoryId) override;
+
+    //! Create a clone.
+    TCategoryIdMapperUPtr clone() const override;
+
+    //! Persist the mapper passing information to \p inserter.
+    void acceptPersistInserter(core::CStatePersistInserter& inserter) const override;
+
+    //! Restore the mapper reading state from \p traverser.
+    bool acceptRestoreTraverser(core::CStateRestoreTraverser& traverser) override;
+
+private:
+    //! Key specifiers for the multi-index
+    using TGlobalCategoryIdVec = std::vector<CGlobalCategoryId>;
+    using TStrGlobalCategoryIdVecMap = std::map<std::string, TGlobalCategoryIdVec>;
```

Comment:
One thing I noticed with the `TStrStrUMap` used to pass around the fields in the input records is that because it's generally small the retrieval time is dominated by calculation of hash values https://github.com/elastic/ml-cpp/blob/9146cdaa6e9d37a3e6b7b3f62fa680e3df24329a/lib/api/CAnomalyJob.cc#L163

This led to workarounds like pre-computing the hashes in some situations https://github.com/elastic/ml-cpp/blob/9146cdaa6e9d37a3e6b7b3f62fa680e3df24329a/include/api/COutputHandler.h#L103

I think there is a cutover point, maybe around 20-30 elements, below which a tree map keyed on strings is faster than a hash map.  (Obviously for numeric keys it's different as that hash is a trivial numeric conversion.)

Per-partition categorization is intended to be used with a relatively low number of partitions.  (The intended use case is several types of logs all going into the same Elasticsearch index, and then a job categorizes the different types of logs separately.)  So that's likely to be relatively few partition field values (i.e. relatively few different types of log files).

So while it's true that with thousands of partition field values a tree map is going to be slower than a hash map, in the expected use case it's going to be faster, or at least not significantly slower.

If someone tries per-partition categorization with thousands of partitions they will quickly run out of memory, as the categorization data structures are pretty big compared to a single anomaly detector.

As you said originally, it would be possible to avoid these lookups altogether by moving the local to global mapping into the low level categorizers in the model library.  I didn't want to pollute the model library classes with this complexity, but I suppose the alternative would be to create a new API library low level categorizer class that inherits from the model library low level categorizer class but does local to global mapping.  Then the model library classes can be tested independently without the extra complexity, but the api library derived class can hold the more advanced functionality that naturally goes with the lower level categorizer.

Reply:
I don't think we need to make this change at this point. Let's just wait and see on performance and if we decide we need to improve performance, we know we can elide these lookups altogether. Basically, I'm happy with this explanation and leaving as a `std::map`.

Reply:
I think I'll have a go at creating a new API library low level categorizer class and see if it makes the overall code nicer.  I think it might because then some of the chunks of code in `CFieldDataCategorizer` that look out of place because there's too much back and forth to the model library classes can move into the new class.  I'll give up if that way of doing things turns out to be worse after looking into it more closely.

Reply:
The refactor I added in 69cc35c9a62a8e76922acbf2212f9b822038de25 avoids looking up the same partition field value twice - once to find the categorizer and once to find the local to global ID mapper - because both are now stored together on a new class of object, `api::CGlobalIdDataCategorizer`.

This extra level of indirection also allows https://github.com/elastic/ml-cpp/pull/1293#discussion_r435511894 to be addressed - since it's in the `api` library the loop body can move to the new class.

Reply:
I'll take another look.